### ISSUE TITLE
make numpy an optional requirement

### ISSUE THREAD

**Issue by ritchie46** [pola-rs/polars#3571]
We should be able to do without numpy.

**Comment by stinodego** [comment]
I'll look into this - seems like a fun one 😃

**Comment by ritchie46** [comment]
Awesome! My idea is that we check if numpy is installed similar to what we do with pyarrow. So uf we have numpy array as arguments we can still dispatch to those constructors.

But for python list inputs we must use our constructors. With regard to type, I think we can do the first non null strategy, unless the user explicitly requests an object datatype.

That strategy is used throughout the codebase so that keeps that consistent and fast.

**Comment by stinodego** [comment]
I think the constructors are quite easy to fix. However, there's quite some functionality that depends on numpy, e.g. `Expr.sin`, `Expr.cos`, `Series.power`, etc.

How would you prefer to handle this? Should I just raise an `ImportError` here, or did you have something else in mind?

```python
    def cos(self) -> "Expr":
        if not _NUMPY_AVAILABLE:
            raise ImportError("'numpy' is required for this functionality.")
        return np.cos(self)  # type: ignore
```

**Comment by ritchie46** [comment]
Yes, that's best for now. Later we could add them natively as expressions.

**Comment by stinodego** [comment]
I think I have most of the cases fixed, but there are a few tough ones in `series.py`. I would appreciate your input on these:

* [`__setitem__`](https://github.com/pola-rs/polars/blob/9e1ab0ddea77e20cb84045fd21368060bf1d3c9a/py-polars/polars/internals/series.py#L514): Here numpy is used to create an array, even if the input is a list or tuple.
* [`take`](https://github.com/pola-rs/polars/blob/9e1ab0ddea77e20cb84045fd21368060bf1d3c9a/py-polars/polars/internals/series.py#L1549): Here numpy is used to convert a list to a numpy array.
* [`set_at_idx`](https://github.com/pola-rs/polars/blob/9e1ab0ddea77e20cb84045fd21368060bf1d3c9a/py-polars/polars/internals/series.py#L2370): Here numpy is used to create an array, even if the input is a list or tuple.

I am not 100% sure on why these conversions happen, and what the alternative would be. Any ideas?

**Comment by ritchie46** [comment]
Archaic reasons that should be changed even if we did keep numpy as a dependency. Can we raise an error atm and open separate issues for those?

I will see if I can do one in a way that better fits current approaches.

P.S. I leave to a festival now, so I will be slow to respond. 😁

**Comment by stinodego** [comment]
> Can we raise an error atm and open separate issues for those?

I've done this for now. Will create the followup issue once the PR is merged.
